Mathlib/GroupTheory/ArchimedeanDensely.lean

```diff
@@ -297,3 +298,135 @@ lemma LinearOrderedCommGroupWithZero.discrete_iff_not_denselyOrdered (G : Type*)
+    Set.WellFoundedOn {x : G | x ≤ g} (· > ·) ↔ Nonempty (G ≃+o ℤ) := by
+  rw [← wellFoundedOn_setOf_le_lt_iff_nonempty_discrete (g := -g)]
+  refine ⟨fun h ↦ (h.mapsTo (- ·) ?_).mono' ?_, fun h ↦ (h.mapsTo (- ·) ?_).mono' ?_⟩ <;>
+  · intro
+    simp [Function.onFun, neg_le]
+
+@[to_additive existing]
+lemma LinearOrderedCommGroup.wellFoundedOn_setOf_le_lt_iff_nonempty_discrete
+    {G : Type*} [LinearOrderedCommGroup G] [Nontrivial G] {g : G} :
+    Set.WellFoundedOn {x : G | g ≤ x} (· < ·) ↔ Nonempty (G ≃*o Multiplicative ℤ) := by
+  let e : G ≃o Additive G := OrderIso.refl G
+  suffices Set.WellFoundedOn {x : G | g ≤ x} (· < ·) ↔ Set.WellFoundedOn {x | e g ≤ x} (· < ·) by
+    rw [this, LinearOrderedAddCommGroup.wellFoundedOn_setOf_le_lt_iff_nonempty_discrete]
+    refine Nonempty.congr (fun f ↦ ?_) (fun f ↦ ?_)
+    · exact ⟨AddEquiv.toMultiplicative' f, by simp⟩
+    · exact ⟨MulEquiv.toAdditive' f, by simp⟩
+  refine ⟨fun h ↦ (h.mapsTo e.symm fun _ ↦ e.le_symm_apply.mpr).mono' ?_,
+    fun h ↦ (h.mapsTo e fun _ ↦ ?_).mono' ?_⟩ <;>
+  simp [Function.onFun]
+
+@[to_additive existing]
+lemma LinearOrderedCommGroup.wellFoundedOn_setOf_ge_gt_iff_nonempty_discrete
+    {G : Type*} [LinearOrderedCommGroup G] [Nontrivial G] (g : G) :
+    Set.WellFoundedOn {x : G | x ≤ g} (· > ·) ↔ Nonempty (G ≃*o Multiplicative ℤ) := by
+  rw [← wellFoundedOn_setOf_le_lt_iff_nonempty_discrete (g := g⁻¹)]
+  refine ⟨fun h ↦ (h.mapsTo (·⁻¹) ?_).mono' ?_, fun h ↦ (h.mapsTo (·⁻¹) ?_).mono' ?_⟩ <;>
+  · intro
+    simp [Function.onFun, inv_le']
+
+lemma LinearOrderedCommGroupWithZero.wellFoundedOn_setOf_le_lt_iff_nonempty_discrete_of_ne_zero
+    {G₀ : Type*} [LinearOrderedCommGroupWithZero G₀] [Nontrivial G₀ˣ] {g : G₀} (hg : g ≠ 0) :
+    Set.WellFoundedOn {x : G₀ | g ≤ x} (· < ·) ↔ Nonempty (G₀ ≃*o ℤₘ₀) := by
+  suffices Set.WellFoundedOn {x : G₀ | g ≤ x} (· < ·) ↔
+    Set.WellFoundedOn {x : G₀ˣ | Units.mk0 g hg ≤ x} (· < ·) by
+    rw [this, LinearOrderedCommGroup.wellFoundedOn_setOf_le_lt_iff_nonempty_discrete]
+    refine Nonempty.congr (fun f ↦ ⟨?_, ?_⟩) (fun f ↦ ⟨?_, ?_⟩)
+    · exact WithZero.withZeroUnitsEquiv.symm.trans f.withZero
+    · intro a b
+      rcases eq_or_ne a 0 with rfl|ha
+      · simp [WithZero.withZeroUnitsEquiv, MulEquiv.withZero]
+      rcases eq_or_ne b 0 with rfl|hb
+      · simp [WithZero.withZeroUnitsEquiv, MulEquiv.withZero]
+      simp [WithZero.withZeroUnitsEquiv, MulEquiv.withZero, ha, hb, ← Units.val_le_val]
+    · exact MulEquiv.unzero (WithZero.withZeroUnitsEquiv.trans f)
+    · intros
+      rw [← WithZero.coe_le_coe]
+      simp [WithZero.withZeroUnitsEquiv, MulEquiv.unzero]
+  rw [← Set.wellFoundedOn_sdiff_singleton (a := 0)]
+  refine ⟨fun h ↦ (h.mapsTo Units.val ?_).mono' ?_,
+    fun h ↦ (h.mapsTo ?_ ?_).mono' ?_⟩
+  · intro
+    simp [← Units.val_le_val]
+  · simp [Function.onFun]
+  · exact fun x ↦ if h : x = 0 then 1 else Units.mk0 x h
+  · intro
+    simp (config := {contextual := true}) [← Units.val_le_val]
```

Comment:
```suggestion
    simp +contextual [← Units.val_le_val]
```
Do you evben need the intro, if you do a contextual simp?

Reply:
Yes, needed if to break open `MapsTo`. Instead, I'll add `MapsTo` so the local simp set